Common/cpp/SharedItems/ShareableValue.cpp

```diff
@@ -72,62 +74,74 @@ void ShareableValue::adapt(jsi::Runtime &rt, const jsi::Value &value, ValueType
-        hostFunction = std::make_shared<HostFunctionHandler>(std::make_shared<jsi::Function>(object.asFunction(rt)), rt);
         containsHostFunction = true;
+        valueContainer = std::make_unique<HostFunctionWrapper>(
+          std::make_shared<HostFunctionHandler>(std::make_shared<jsi::Function>(object.asFunction(rt)), rt),
+          &rt
+        );
       } else {
         // a worklet
         type = ValueType::WorkletFunctionType;
-        frozenObject = std::make_shared<FrozenObject>(rt, object, module);
-        containsHostFunction |= frozenObject->containsHostFunction;
+        std::shared_ptr<FrozenObject> frozenObject = std::make_shared<FrozenObject>(rt, object, module);
+        containsHostFunction = frozenObject->containsHostFunction;
+        valueContainer = std::make_unique<FrozenObjectWrapper>(frozenObject);
         if (isRNRuntime && !containsHostFunction) {
           addHiddenProperty(rt, createHost(rt, frozenObject), object, HIDDEN_HOST_OBJECT_PROP);
         }
       }
     } else if (object.isArray(rt)) {
       type = ValueType::ArrayType;
       auto array = object.asArray(rt);
+      std::vector<std::shared_ptr<ShareableValue>> result;
       for (size_t i = 0, size = array.size(rt); i < size; i++) {
         auto sv = adapt(rt, array.getValueAtIndex(rt, i), module);
         containsHostFunction |= sv->containsHostFunction;
-        frozenArray.push_back(sv);
+        result.push_back(sv);
       }
+      valueContainer = std::make_unique<FrozenArrayWrapper>(result);
     } else if (object.isHostObject<MutableValue>(rt)) {
       type = ValueType::MutableValueType;
-      mutableValue = object.getHostObject<MutableValue>(rt);
+      valueContainer = std::make_unique<MutableValueWrapper>(object.getHostObject<MutableValue>(rt));
       adaptCache(rt, value);
     } else if (object.isHostObject<RemoteObject>(rt)) {
       type = ValueType::RemoteObjectType;
-      remoteObject = object.getHostObject<RemoteObject>(rt);
+      valueContainer = std::make_unique<RemoteObjectWrapper>(
+         object.getHostObject<RemoteObject>(rt)
+      );
       adaptCache(rt, value);
     } else if (objectType == ValueType::RemoteObjectType) {
       type = ValueType::RemoteObjectType;
-      remoteObject = std::make_shared<RemoteObject>(rt, object, module, module->scheduler);
+      valueContainer = std::make_unique<RemoteObjectWrapper>(
+        std::make_shared<RemoteObject>(rt, object, module, module->scheduler)
+      );
     } else {
       // create frozen object based on a copy of a given object
       type = ValueType::ObjectType;
-      frozenObject = std::make_shared<FrozenObject>(rt, object, module);
-      containsHostFunction |= frozenObject->containsHostFunction;
+      std::shared_ptr<FrozenObject> frozenObject = std::make_shared<FrozenObject>(rt, object, module);
+      containsHostFunction = frozenObject->containsHostFunction;
+      valueContainer = std::make_unique<FrozenObjectWrapper>(frozenObject);
```

Comment:
same as above: why don't you assign `valueContainer` immediately and why did you get rid of the `or` operator?